### PR TITLE
vic3: validate commander ranks and orders

### PR DESCRIPTION
**Pull request by amtep**

### CODE REVIEW

File: src/vic3/data/commanders.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        vd.field_item("texture", Item::File);
+        vd.field_choice("military_type", &["army", "navy"]);
+        vd.field_choice(
+            "ai_usage",
+            &["advance", "defend", "intercept", "raid_convoys", "escort_convoys"],
+        );
+        vd.field_bool("is_basic_order_type");
+
+        vd.field_validated_key_block("possible", |key, block, data| {
+            // TODO: this should have an `error_check` field somewhere inside
```
Comment (my4ng) [inline src/vic3/data/commanders.rs:49]: I think all should at least have a general/admiral invalid error_check, maybe as a tip

File: src/vic3/data/commanders.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            // TODO: `NOT = { has_trait = foo }` in this trigger will be misrepresented in the UI
+            let mut sc = ScopeContext::new(Scopes::Character, key);
+            validate_trigger(block, data, &mut sc, Tooltipped::No);
+        });
+        vd.field_validated_block("modifier", |block, data| {
+            let vd = Validator::new(block, data);
+            validate_modifs(block, data, ModifKinds::all(), vd);
```
Comment (my4ng) [inline src/vic3/data/commanders.rs:56]: Should be Battle | Character | Unit

File: src/vic3/data/commanders.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            vd.ban_field("naval_entity", || "navy");
+        }
+
+        vd.field_numeric_range("indicator_position_angle", 0.0, 360.0);
+        vd.field_numeric_range("indicator_position_angle_for_enemy", 0.0, 360.0);
+        vd.field_item("clicksound", Item::Sound);
+        vd.field_numeric("experience");
+
+        let mut sc = ScopeContext::new(Scopes::Character, key);
+        vd.field_script_value("ai_weight", &mut sc);
```
Comment (my4ng) [inline src/vic3/data/commanders.rs:75]: Has the sc being rooted to character been verified?

File: src/vic3/data/commanders.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            validate_trigger(block, data, &mut sc, Tooltipped::No);
+        });
+        vd.field_validated_block("modifier", |block, data| {
+            let vd = Validator::new(block, data);
+            validate_modifs(block, data, ModifKinds::all(), vd);
+        });
+
+        // TODO: this should be a state inside a unit's Item::Entity
+        // The connection between the units and the entities is very indirect.
+        vd.field_value("entity_animation");
```
Comment (my4ng) [inline src/vic3/data/commanders.rs:61]: Yes each unit's entity has a state with the given name, e.g. `state = { name = "defend" `.

File: src/vic3/data/commanders.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        // but that doesn't limit the possible `rank_value` values, since they can be set in other
+        // ways than promotion.
+        vd.field_integer_min("rank_value", 1);
+
+        for field in
+            &["character_modifier", "general_modifier", "admiral_modifier", "country_modifier"]
+        {
+            vd.field_validated_block(field, |block, data| {
+                let vd = Validator::new(block, data);
+                validate_modifs(block, data, ModifKinds::all(), vd);
```
Comment (my4ng) [inline src/vic3/data/commanders.rs:112]: should be Character for first three and Country for the last?

**Review by my4ng — CHANGES_REQUESTED** [review]

File: src/validator.rs [hunk 6]
```diff
@@ -471,6 +471,25 @@ impl<'a> Validator<'a> {
         });
     }
 
+    /// Expect field `name`, if present, to be set to an integer at least `min`
+    /// Expect no more than one `name` field.
+    /// Returns true iff the field is present.
+    #[cfg(feature = "vic3")]
+    pub fn field_integer_min(&mut self, name: &str, min: i64) {
```
[the remark this answers is not in the document]
Comment (amtep) [reply src/validator.rs:478]: That's a good idea

File: src/vic3/data/commanders.rs (hunk 1 above)
Comment (amtep) [reply src/vic3/data/commanders.rs:49]: Yeah it's just that the example has the error_check buried in an AND :)

File: src/vic3/data/commanders.rs (hunk 5 above)
Comment (amtep) [reply src/vic3/data/commanders.rs:112]: No, characters and countries can have all sorts of modifierkinds in vic3.

Like a character can have an interest group modifier that's used if that character is an interest group leader.

File: src/vic3/data/commanders.rs (hunk 3 above)
Comment (amtep) [reply src/vic3/data/commanders.rs]: I don't know how to test it. I didn't get any error messages out of breaking it.

**Review by my4ng — APPROVED** [review]

**Comment by amtep** [comment]
I did add the undocumented comment and an `error_check` check :)  Probably you didn't notice because of my sneaky rebase.

**Pull request merged**